sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/estimation/JoinEstimation.scala

```diff
@@ -0,0 +1,175 @@
+        // cartesian product.
+        leftRows * rightRows
+      } else {
+        ceil(BigDecimal(leftRows * rightRows) / BigDecimal(ndvDenom))
+      }
+
+      // 3. Update statistics based on the output of join
+      val originalStats = leftStats.colStats ++ rightStats.colStats
+      val outputWithStat = join.output.filter(attr => originalStats.contains(attr.name))
+
+      val outputColStats = new mutable.HashMap[String, ColumnStat]()
+      if (ndvDenom == 0) {
+        // empty output
+        outputWithStat.foreach(a => outputColStats.put(a.name, emptyColumnStat(a.dataType)))
+      } else if (ndvDenom < 0) {
+        // cartesian product, column stats will not change
+        outputWithStat.foreach(a => outputColStats.put(a.name, originalStats(a.name)))
+      } else {
+        val leftRatio = BigDecimal(outputRows) / BigDecimal(leftRows)
+        val rightRatio = BigDecimal(outputRows) / BigDecimal(rightRows)
+        outputWithStat.foreach { a =>
+          // check if this attribute is a join key
+          if (intersectedStats.contains(a.name)) {
+            outputColStats.put(a.name, intersectedStats(a.name))
+          } else {
+            val oldColStat = originalStats(a.name)
+            val oldNdv = oldColStat.distinctCount
+            // We only change (scale down) the number of distinct values if the number of rows
+            // decreases after join, because join won't produce new values even if the number of
+            // rows increases.
+            val newNdv = if (left.outputSet.contains(a) && leftRatio < 1) {
+              ceil(BigDecimal(oldNdv) * leftRatio)
+            } else if (right.outputSet.contains(a) && rightRatio < 1) {
+              ceil(BigDecimal(oldNdv) * rightRatio)
+            } else {
+              oldNdv
+            }
+            outputColStats.put(a.name, oldColStat.copy(distinctCount = newNdv, nullCount = 0))
+          }
+        }
+      }
+
+      Some(Statistics(
+        sizeInBytes = outputRows * getRowSize(join.output, outputColStats.toMap),
+        rowCount = Some(outputRows),
+        colStats = outputColStats.toMap,
+        isBroadcastable = false))
+
+    case _ => None
+  }
+
+  def emptyColumnStat(dataType: DataType): ColumnStat = {
+    ColumnStat(distinctCount = 0, min = None, max = None, nullCount = 0,
+      avgLen = dataType.defaultSize, maxLen = dataType.defaultSize)
+  }
+
+  def intersectedColumnStat(
+      origin: ColumnStat,
+      newDistinctCount: BigInt,
+      newMin: Option[Any],
+      newMax: Option[Any]): ColumnStat = {
+    origin.copy(distinctCount = newDistinctCount, min = newMin, max = newMax, nullCount = 0)
+  }
+
+  def extractJoinKeys(
+      leftKeys: Seq[Expression],
+      rightKeys: Seq[Expression]): Seq[(AttributeReference, AttributeReference)] = {
+    leftKeys.zip(rightKeys).flatMap {
+      case (ExtractAttr(left), ExtractAttr(right)) => Some((left, right))
+      // Currently we don't deal with equal joins like key1 = key2 + 5.
```

Comment:
If we are not ready for expressions, I think we should not handle `Cast` either, as it may be tricky to handle overflow correctly(e.g. cast long to int)